src/modules/sensors/vehicle_imu/Integrator.hpp

```diff
@@ -95,29 +83,106 @@ class Integrator
 	 *
 	 * @param integral_dt	Get the dt in us of the current integration.
 	 * @return		true if integral valid
 	 */
-	bool reset(matrix::Vector3f &integral, uint32_t &integral_dt);
+	bool reset(matrix::Vector3f &integral, uint32_t &integral_dt)
+	{
+		if (integral_ready()) {
+			integral = _alpha;
+			integral_dt = roundf(_integral_dt * 1e6f); // seconds to microseconds
 
-private:
-	uint64_t _last_integration_time{0}; /**< timestamp of the last integration step */
-	uint64_t _last_reset_time{0};       /**< last auto-announcement of integral value */
+			// reset
+			_alpha.zero();
+			_integral_dt = 0;
+			_integrated_samples = 0;
 
-	matrix::Vector3f _alpha{0.f, 0.f, 0.f};            /**< integrated value before coning corrections are applied */
-	matrix::Vector3f _last_alpha{0.f, 0.f, 0.f};       /**< previous value of _alpha */
-	matrix::Vector3f _beta{0.f, 0.f, 0.f};             /**< accumulated coning corrections */
-	matrix::Vector3f _last_val{0.f, 0.f, 0.f};         /**< previous input */
-	matrix::Vector3f _last_delta_alpha{0.f, 0.f, 0.f}; /**< integral from previous previous sampling interval */
+			return true;
+		}
+
+		return false;
+	}
 
-	uint32_t _reset_interval_min{1}; /**< the interval after which the content will be published and the integrator reset */
+protected:
 
-	uint8_t _integrated_samples{0};
+	matrix::Vector3f integrate(const matrix::Vector3f &val, const float dt)
+	{
+		// Use trapezoidal integration to calculate the delta integral
+		_integrated_samples++;
+		_integral_dt += dt;
+		const matrix::Vector3f delta_alpha{(val + _last_val) *dt * 0.5f};
+		_last_val = val;
+
+		return delta_alpha;
+	}
+
+	matrix::Vector3f _alpha{0.f, 0.f, 0.f};    /**< integrated value before coning corrections are applied */
+	matrix::Vector3f _last_val{0.f, 0.f, 0.f}; /**< previous input */
+	float _integral_dt{0};
+
+	float _reset_interval_min{0.005f}; /**< the interval after which the content will be published and the integrator reset */
 	uint8_t _reset_samples_min{1};
 
-	const bool _coning_comp_on{false};                       /**< true to turn on coning corrections */
+	uint8_t _integrated_samples{0};
+};
+
+class IntegratorConing : public Integrator
+{
+public:
+	IntegratorConing() = default;
+	~IntegratorConing() = default;
+
+	/**
+	 * Put an item into the integral.
+	 *
+	 * @param timestamp	Timestamp of the current value.
+	 * @param val		Item to put.
+	 * @return		true if data was accepted and integrated.
+	 */
+	void put(const matrix::Vector3f &val, const float dt)
+	{
+		// Use trapezoidal integration to calculate the delta integral
+		const matrix::Vector3f delta_alpha{integrate(val, dt)};
+
+		// Calculate coning corrections
+		// Coning compensation derived by Paul Riseborough and Jonathan Challinger,
+		// following:
+		// Tian et al (2010) Three-loop Integration of GPS and Strapdown INS with Coning and Sculling Compensation
+		// Sourced: http://www.sage.unsw.edu.au/snap/publications/tian_etal2010b.pdf
```

Comment:
This reference is really difficult (impossible?) to find. Should we change it with https://arc.aiaa.org/doi/pdf/10.2514/2.4228 that contains the same equations?

Reply:
Might as well. FYI @priseborough

Reply:
Done.